core/trino-main/src/main/java/io/trino/operator/scalar/MathFunctions.java

```diff
@@ -858,6 +864,11 @@ public static long roundReal(@SqlType(StandardTypes.REAL) long num, @SqlType(Sta
         if (rescaledRound != Long.MAX_VALUE) {
             result = sign * (rescaledRound / factor);
         }
+        else if (rescaled == Double.POSITIVE_INFINITY) {
+            double positive = sign * numInFloat;
+            BigDecimal bd = BigDecimal.valueOf(positive);
+            result = sign * bd.setScale((int) decimals, RoundingMode.HALF_UP).doubleValue();
```

Comment:
Having tested this locally, I can say that the performance of this approach is _terrible_. The unit test examples added take over 5 seconds to perform each rounding operation.

I wonder if there's something better we can do when handling high precision rounding. Note that `BigDecimal.valueOf(double)` is implemented as:
```
public static BigDecimal valueOf(double val) {
        // <irrelevant comment ellided>
        return new BigDecimal(Double.toString(val));
    }
```

where `Double.toString` internally is already doing a _lot_ of the same work we're trying to accomplish in our rounding routine in order to produce the string output which is then just parsed back in. There might be a more efficient approach we can take from the implementation there to get a correct result without such a large performance hit.

Reply:
Yes, I totally agree. I tested the performance locally as well! I think the performance implication here is the main reason why we use rescaled approach for other cases, given that BigDecimal approach should work for any input. I am open to other ideas how we can handle this edge case with better performance than BigDecimal approach.